In [ ]:
!pip install pyspark

In [ ]:
!pip install surprise

In [ ]:
from pyspark.sql import SparkSession
import numpy as np
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
spark = SparkSession.builder.master("local[*]").getOrCreate()

In [ ]:
df1=pd.read_csv('/content/drive/MyDrive/Colab/BigData/data/midterm/tmdb_5000_credits.csv')
df2=pd.read_csv('/content/drive/MyDrive/Colab/BigData/data/midterm/tmdb_5000_movies.csv')

In [ ]:
df1.head(3)

movie_id  ...                                               crew
0     19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1       285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2    206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...

[3 rows x 4 columns]

In [ ]:
df2.head(3)

budget  ... vote_count
0  237000000  ...      11800
1  300000000  ...       4500
2  245000000  ...       4466

[3 rows x 20 columns]

In [ ]:
df1.columns = ['id','tittle','cast','crew']
df2= df2.merge(df1,on='id')
df2.head(3)

budget  ...                                               crew
0  237000000  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1  300000000  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2  245000000  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...

[3 rows x 29 columns]

# **Lọc nhân khẩu - Demographic Filtering**
  Trước tiên cần tính rating trung bình của từng bộ phim, sau đó sắp xếp và giới thiệu các bộ phim có rating tốt nhất cho người dùng
  Dùng trung bình cộng sẽ không chính xác, trong trường hợp này sẽ dùng weighted rating

  ![](https://image.ibb.co/jYWZp9/wr.png)


*   v: số vote của từng bộ phim
*   m: số vote tối thiểu để được giới thiệu đến người dùng
*   R: điểm rating trung bình của bộ phim
*   C: rating trung bình trên toàn bộ tập dataset







In [ ]:
C= df2['vote_average'].mean()
C

6.092171559442011

điều này cho thấy rating trung bình cho toàn bộ phim trong tập dataset là từ 6 trở lên

Xác định số phiếu tối thiểu để bộ phim được đưa vào list recommend đến người dùng. Ở đây, chỉ số quantile = 0.8 => Nghĩa là bộ phim phải có số vote nhiều hơn 80% bộ phim trong tập dataset

In [ ]:
m= df2['vote_count'].quantile(0.8)
m

957.6000000000004

In [ ]:
q_movies = df2.copy().loc[df2['vote_count'] >= m]
q_movies.shape

(961, 29)

Từ kết quả trên, có thể thấy có 961 bộ phim thoả mãn điều kiện đã cho. Bước tiếp theo sẽ tính weighted rating để đưa ra list các bộ phim đứng top và recommend cho người dùng

In [ ]:
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [ ]:
q_movies['score'] = q_movies.apply(weighted_rating, axis=1)

In [ ]:
q_movies = q_movies.sort_values('score', ascending=False)

q_movies[['title', 'vote_count', 'vote_average', 'score']].head(10)

title  vote_count  vote_average     score
1881  The Shawshank Redemption        8205           8.5  8.248353
662                 Fight Club        9413           8.3  8.096134
3337             The Godfather        5893           8.4  8.077404
3232              Pulp Fiction        8428           8.3  8.074738
65             The Dark Knight       12002           8.2  8.044250
809               Forrest Gump        7927           8.2  7.972814
96                   Inception       13752           8.1  7.969290
95                Interstellar       10867           8.1  7.937399
1990   The Empire Strikes Back        5879           8.2  7.904757
1818          Schindler's List        4329           8.3  7.900080

# **Content-Based system**
Trong tập dataset này, chúng ta sẽ sử dụng atrribute overview để tìm điểm tương đồng giữa các bộ phim, từ đó đưa ra đề xuất

In [ ]:
df2['overview'].head(5)

0    In the 22nd century, a paraplegic Marine is di...
1    Captain Barbossa, long believed to be dead, ha...
2    A cryptic message from Bond’s past sends him o...
3    Following the death of District Attorney Harve...
4    John Carter is a war-weary, former military ca...
Name: overview, dtype: object

Điều đầu tiên cần làm là chuyển về dạng vector từ (word)

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [ ]:
#remove stop word trong tiếng anh như 'the', 'a', ...
tfidf = TfidfVectorizer(stop_words='english')

df2['overview'] = df2['overview'].fillna('')

tfidf_matrix = tfidf.fit_transform(df2['overview'])

tfidf_matrix.shape

(4803, 20978)

Bước tiếp theo là tìm điểm tương tự
Bài này sẽ được triển khai theo độ tương tự cosine (cosine similarity)

In [ ]:
from sklearn.metrics.pairwise import linear_kernel
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

In [ ]:
indices = pd.Series(df2.index, index=df2['title']).drop_duplicates()

In [ ]:
def get_recommendations(title, cosine_sim=cosine_sim):

    idx = indices[title]
    sim_scores = list(enumerate(cosine_sim[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:11]
    movie_indices = [i[0] for i in sim_scores]

    return df2['title'].iloc[movie_indices]

In [ ]:
get_recommendations('Spectre')

1343    Never Say Never Again
4071    From Russia with Love
3162              Thunderball
1717               Safe Haven
11          Quantum of Solace
4339                   Dr. No
29                    Skyfall
1880              Dance Flick
3336     Diamonds Are Forever
1743                Octopussy
Name: title, dtype: object

# **Collaborative Filtering**

In [ ]:
from surprise import Reader, Dataset, SVD
from surprise.model_selection import cross_validate
reader = Reader()
ratings = pd.read_csv('/content/drive/MyDrive/Colab/BigData/data/midterm/ratings_small.csv')
ratings.head()

userId  movieId  rating   timestamp
0       1       31     2.5  1260759144
1       1     1029     3.0  1260759179
2       1     1061     3.0  1260759182
3       1     1129     2.0  1260759185
4       1     1172     4.0  1260759205

In [ ]:
data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)
algo = SVD()
cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8908  0.8944  0.9060  0.8933  0.8955  0.8960  0.0052  
MAE (testset)     0.6879  0.6873  0.6966  0.6894  0.6857  0.6894  0.0038  
Fit time          4.70    4.69    4.67    4.69    4.67    4.68    0.01    
Test time         0.29    0.16    0.25    0.15    0.18    0.21    0.05    


{'fit_time': (4.699723243713379,
  4.689619779586792,
  4.6720921993255615,
  4.686875820159912,
  4.673409938812256),
 'test_mae': array([0.687854  , 0.68732205, 0.69656632, 0.68943275, 0.68574352]),
 'test_rmse': array([0.8908195 , 0.89438199, 0.90600567, 0.89329884, 0.8954767 ]),
 'test_time': (0.2853515148162842,
  0.1554121971130371,
  0.25430774688720703,
  0.1545724868774414,
  0.18138766288757324)}

In [ ]:
trainset = data.build_full_trainset()
algo.fit(trainset)

In [ ]:
ratings[ratings['userId'] == 1]

userId  movieId  rating   timestamp
0        1       31     2.5  1260759144
1        1     1029     3.0  1260759179
2        1     1061     3.0  1260759182
3        1     1129     2.0  1260759185
4        1     1172     4.0  1260759205
5        1     1263     2.0  1260759151
6        1     1287     2.0  1260759187
7        1     1293     2.0  1260759148
8        1     1339     3.5  1260759125
9        1     1343     2.0  1260759131
10       1     1371     2.5  1260759135
11       1     1405     1.0  1260759203
12       1     1953     4.0  1260759191
13       1     2105     4.0  1260759139
14       1     2150     3.0  1260759194
15       1     2193     2.0  1260759198
16       1     2294     2.0  1260759108
17       1     2455     2.5  1260759113
18       1     2968     1.0  1260759200
19       1     3671     3.0  1260759117

In [ ]:
algo.predict(1, 302, 3)

Prediction(uid=1, iid=302, r_ui=3, est=2.796047377328743, details={'was_impossible': False})